flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/DataStream.java

```diff
@@ -1194,6 +1197,37 @@ public ExecutionConfig getExecutionConfig() {
 		return returnStream;
 	}
 
+	/**
+	 * Method for passing user defined operators along with the type information that will transform the DataStream.
+	 *
+	 * @param operatorName name of the operator, for logging purposes
+	 * @param outTypeInfo the output type of the operator
+	 * @param operatorFactory the factory for the operator.
+	 * @param <R> type of the return stream
+	 * @return the data stream constructed
+	 */
+	@PublicEvolving
+	public <R> SingleOutputStreamOperator<R> transform(String operatorName, TypeInformation<R> outTypeInfo,
+			OneInputStreamOperatorFactory<T, R> operatorFactory) {
```

Comment:
Continuing my comment about deprecating the previous method:
imo, providing this method as the new alternative maybe not so user friendly as users would need create additional layer (factory) to provide their own operators.
Also, it may be not so easy to provide api stability from runtime point of view, as the operator factories were mostly internal so far.

Maybe mark this method as `@Internal` at least for now? **Note** that `StreamOperatorFactory` (parent of `OneInputStreamOperatorFactory`) is also marked as `@Internal`.

**Nit:** the documentation string of this method is not exact, as the users don't pass an operator instance.

Reply:
I left it as PublicEvolving but added a comment that the other method is preferred.

Reply:
It makes it less convenient for power users, but it's necessary for cleaner API and runtime code. I would mark as deprecated, because I would like to have at least flink built-in operators to move away from this old API in favour of `StreamOperatorFactory` as soon as possible.

But ultimately we can merge it with or without `@Deprecated` note and question would be more towards the @aljoscha what does he prefer.

Reply:
Aljoscha is fine with the current state.